### PR TITLE
[RuntimeEnv] Raise RuntimeEnvSetupError when Actor Creation Failed due to It

### PR DESCRIPTION
**Pull request by lixin-wei**
## Why are these changes needed?

1. Now RuntimeEnvSetupError will be raised if actor is failed due to runtime env.
2. Refactor actor death code path. Abstract `ActorDeathCause`. Unifying actor creation failure and runtime env failure together.
```protobuf
message ActorDeathCause {
  oneof context {
    CreationTaskFailureContext creation_task_failure_context = 1;
    RuntimeEnvSetupFailureContext runtime_env_setup_failure_context = 2;
  }
}

// Indicates that this actor is marked as DEAD due to actor creation task failure.
message CreationTaskFailureContext {
  // The exception thrown in creation task.
  RayException creation_task_exception = 1;
}

//Indicates that this actor is marked as DEAD due to runtime environment setup failure.
message RuntimeEnvSetupFailureContext {
  // TODO(sang,lixin) Get this error message from agent.
  string error_message = 1;
}
```
## Related issue number

https://github.com/ray-project/ray/issues/19514, https://github.com/ray-project/ray/issues/19824

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by lixin-wei** [comment]
Hi @rkooo567, the failed CI seems not related. Would you please have a look?

**Comment by rkooo567** [comment]
Yeah I will take a look at it on Monday!

File: src/ray/gcs/gcs_server/gcs_actor_scheduler.cc
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_actor_scheduler.cc:504]: Can we just not write a method for this? I think it is easier to read that it is in here (we don't need to follow the function def and this method is not resuable)

File: src/ray/gcs/gcs_server/gcs_actor_scheduler.cc
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_actor_scheduler.cc:511]: Can you remove this comment?

Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_actor_scheduler.cc:511]: Looks like you copied and pasted something by mistake

File: src/ray/protobuf/gcs.proto
Comment (rkooo567) [inline src/ray/protobuf/gcs.proto:166]: Instead of having this bool, can we create a new message like `message ActorDeadCause`? And within the dead cause, we can use `runtime_env_setup_failed ` as an enum. This will help us adding more observability into actor death in the future (we can add more dead states other than `runtime_env_setup_failed ` this way)

**Comment by lixin-wei** [comment]
@rkooo567 Thank you for your review! I'll ping you when I'm finished.

File: src/ray/gcs/gcs_server/gcs_actor_scheduler.cc
Comment (lixin-wei) [reply src/ray/gcs/gcs_server/gcs_actor_scheduler.cc:504]: I abstract this function because this part of code is duplicate in `GcsBasedActorScheduler` and `RayletBasedActorScheduler`

File: src/ray/protobuf/gcs.proto
Comment (lixin-wei) [reply src/ray/protobuf/gcs.proto:166]: I agree with you, I'll add one.

**Comment by rkooo567** [comment]
@lixin-wei Lmk when it is ready! I have one task I might be relying on this PR (that's improving the actor failure message)

**Comment by lixin-wei** [comment]
> @lixin-wei Lmk when it is ready! I have one task I might be relying on this PR (that's improving the actor failure message)

OK, I'll finish it in this week!

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc
Comment (lixin-wei) [reply src/ray/gcs/gcs_server/gcs_actor_manager.cc:935]: `DestroyActor` is a public method used to kill actors, it can be called at any time. No matter actor is created or not. It'll do additional cleanups to that actor.

`ReconstructActor (need_reschedule=false)` is a private method, it can only be called after actor is created. Or some data won't be cleanup.

**Comment by lixin-wei** [comment]
@rkooo567 I'm working on the CI, but the code is ready to be reviewed. Could you have a look?

The main change is I added an ActorDeathCause:
```
message ActorDeathCause {
  oneof context {
    CreationTaskFailureContext creation_task_failure_context = 1;
    RuntimeEnvSetupFailureContext runtime_env_setup_failure_context = 2;
  }
}

// Indicates that this actor is marked as DEAD due to actor creation task failure.
message CreationTaskFailureContext {
  // The exception thrown in creation task.
  RayException creation_task_exception = 1;
}

//Indicates that this actor is marked as DEAD due to runtime environment setup failure.
message RuntimeEnvSetupFailureContext {
  // TODO(sang,lixin) Get this error message from agent.
  string error_message = 1;
}
```

**Comment by rkooo567** [comment]
Taking a look today

File: src/ray/protobuf/gcs.proto
Comment (rkooo567) [inline src/ray/protobuf/gcs.proto:99]: Why don't we rename it to ActorInitFailure? Technically runtime env setup failure also is creation task failure.

File: src/ray/protobuf/gcs.proto
Comment (rkooo567) [inline src/ray/protobuf/gcs.proto]:
```suggestion
// Indicates that this actor is marked as DEAD due to runtime environment setup failure.
```

Space!

**Review by rkooo567 — COMMENTED** [review]
Looks pretty good! Mostly about nit comments. I think we can merge it in the next iteration...

File: src/ray/core_worker/actor_manager.cc
Comment (lixin-wei) [reply src/ray/core_worker/actor_manager.cc:223]: Sadly no... For OneOf cases it doesn't have such methods...

Comment (rkooo567) [reply src/ray/core_worker/actor_manager.cc:223]: Oh it is not enum.. I see..

Comment (rkooo567) [reply src/ray/core_worker/actor_manager.cc:223]: Maybe we can require them to have to_string method per each context?

Comment (rkooo567) [reply src/ray/core_worker/actor_manager.cc:223]: Hmm I'd say it is up to you... It's definitely nice to have better error message, but it has a bit more complexity

File: python/ray/_raylet.pyx
Comment (lixin-wei) [reply python/ray/_raylet.pyx:732]: Just leave it alone? I think the more comment, the better haha. So as your other comments about the comment.

File: src/ray/core_worker/task_manager.cc
Comment (lixin-wei) [reply src/ray/core_worker/task_manager.cc:523]: I agree with you, but it seems out of the scope of this PR. I'll open another PR to do this. Here I just add more comments about this part of code when debugging through it.

File: python/ray/_raylet.pyx
Comment (rkooo567) [reply python/ray/_raylet.pyx:732]: I don't think more comments == better. Comment can be easily diverged from implementation because people don't usually fix it when they change code, so we should avoid having redundant comments (if we have a chance).

File: src/ray/core_worker/task_manager.cc
Comment (rkooo567) [reply src/ray/core_worker/task_manager.cc:523]: This sounds good to me!

File: src/ray/protobuf/gcs.proto
Comment (lixin-wei) [reply src/ray/protobuf/gcs.proto:99]: No, creation task is executed after the runtime env is ready. Runtime env setup is in the worker process creation step.

File: python/ray/_raylet.pyx
Comment (rkooo567) [reply python/ray/_raylet.pyx:732]: Same for the below one. If somebody changes the return type of `to_bytes`, it will just add cognitive overhead (since the comment and code are now different), which is a tech debt we should avoid as much as possible.

Comment (lixin-wei) [reply python/ray/_raylet.pyx:732]: OK, I agree with you. I'll remove them.

**Comment by rkooo567** [comment]
Is this ready?

**Comment by lixin-wei** [comment]
> Is this ready (did you address the last set of feedback)?

Oh sorry, not yet. I'll resolve all comments today!

File: src/ray/core_worker/transport/direct_actor_transport.cc
Comment (lixin-wei) [reply src/ray/core_worker/transport/direct_actor_transport.cc:127]: It's a const reference in PendingTaskFailed. I think it doesn't need to be moved.

File: src/ray/core_worker/transport/direct_actor_transport.cc
Comment (lixin-wei) [reply src/ray/core_worker/transport/direct_actor_transport.cc:115]: Why unique ptr here? I think we allow this ptr to be copied in the following codes, like PendingTaskFailed(currently won't be copied, but I think we should allow this in case of changing interface in the future.).

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc
Comment (lixin-wei) [reply src/ray/gcs/gcs_server/gcs_actor_manager.cc:946]: But some cleanup is already done when the actor is created. We don't need to clean it again. ReconstructActor is good enough here.

File: src/ray/core_worker/actor_manager.cc
Comment (lixin-wei) [reply src/ray/core_worker/actor_manager.cc:223]: It seems not easy to implement, which will increase this PR's size by a lot. Let me do this in a follow-up PR.

File: src/ray/gcs/gcs_server/gcs_actor_manager.h
Comment (lixin-wei) [reply src/ray/gcs/gcs_server/gcs_actor_manager.h:316]: It' a scheduling failure but not an actor death failure. In this function maybe actor might not be marked as dead. So I think we shouldn't pass death_cause here.

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc
Comment (lixin-wei) [reply src/ray/gcs/gcs_server/gcs_actor_manager.cc:833]: I have some questions. Let me discuss it in slack with you :)

**Comment by lixin-wei** [comment]
@rkooo567 All comments resolved/replied :)

File: src/ray/core_worker/transport/direct_actor_transport.cc
Comment (rkooo567) [reply src/ray/core_worker/transport/direct_actor_transport.cc:115]: Shared pointer means shared ownership, and it doesn’t really make sense in this function since the function will lose ownership right away. We have a code guideline to make ownership more clear, so I think unique pointer or optional is probably a better choice here. Shared pointer also has unnecessary atomic operation under the hood when returned.

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc
Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_actor_manager.cc:946]: Isn’t DestroyActor already idempotent? 

I got your point, I think when reading code, this can easily cause confusion (like why not DestroyActor). When you described the high level function, it sounds easier to understand to use DestroyActor because we don’t need to understand if actor data has cleaned up or not at this point

File: src/ray/gcs/gcs_server/gcs_actor_manager.h
Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_actor_manager.h:316]: Hmm I personally don’t like the argument here because it exposes the implementation detail. Is this function only called when scheduling has failed?

**Review by rkooo567 — APPROVED** [review]
Lgtm! Thanks for the pr!

File: src/ray/gcs/gcs_server/gcs_actor_manager.h
Comment (lixin-wei) [reply src/ray/gcs/gcs_server/gcs_actor_manager.h:316]: Personally, I don't like it too, haha. But I also don't want to abstract a 'ScheduleFailureCause' here. Do you have better ideas?

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc
Comment (lixin-wei) [reply src/ray/gcs/gcs_server/gcs_actor_manager.cc:946]: OK I agree with you. DestroyActor is easier to understand here. I'll change it.

File: src/ray/gcs/gcs_server/gcs_actor_manager.h
Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_actor_manager.h:316]: I think we can keep it like this for now in that case 😂 sad, but I can let it go haha

**Comment by rkooo567** [comment]
@lixin-wei can you resolve the conflict? I will merge it after that! Maybe we should move quickly (please ping me as soon as tests pass) as Chen is doing some work there...

**Comment by rkooo567** [comment]
@lixin-wei actually, I just fixed it and pushed code.

**Comment by lixin-wei** [comment]
Oh thank you, but wait me for several hours I still have 2 comments to resolve... I'm out now, I'll finish all them before dinner!

**Comment by rkooo567** [comment]
Thanks for doing this in the weekends :)!!!

**Comment by lixin-wei** [comment]
@rkooo567 All done! :)

**Comment by rkooo567** [comment]
Cc @wuisawesome can u review the protobuf real quick? We need ur approval here

**Comment by rkooo567** [comment]
Oh no.. it has merge conflict again.. (due to Chen's refactoring there now).

Really hate that we should wait for approvals for gcs.proto. cc @DmitriGekhtman @wuisawesome can you guys move the autoscaler protobuf to a separate file instead of locking the whole gcs.proto?

**Review by ericl — APPROVED** [review]
approving for proto changes

**Comment by lixin-wei** [comment]
conflict resolved :)

**Comment by rkooo567** [comment]
@lixin-wei Thanks! I was actually doing the same thing now lol.

**Comment by rkooo567** [comment]
I will merge it as soon as tests pass!

**Comment by rkooo567** [comment]
Failing test seems unrelated

**Pull request merged**
